File: Python3/752. Open the Lock.py
class Solution:
    def openLock(self, deadends: List[str], target: str) -> int:
        
        # get the four neighbors of the current node
        def neighbors(node):
            for i in range(4):
                
                # get the integer representation of one of the four digits 
                x = int(node[i])
                
                # increment or decrement it 
                for j in (-1, 1):
                    y = (x+j)%10
                    yield node[:i] + str(y) + node[i+1:]
                    
        # quick check using set
        dead = set(deadends)
        
        # the initial state 
        seen = {'0000'}
        
        # use a queue to do BFS, 0 indicates the current move
        q = deque([('0000', 0)])
        
        while q:
            node, move = q.popleft()
            if node == target:
                return move
            
            # continue searching for other possible ways
            if node in dead:
                continue
                
            for nei in neighbors(node):
                if nei not in seen:
                    seen.add(nei)
                    q.append((nei, move+1))
                    
        return -1
            
        

File: Java/752. Open the Lock.java
class Solution {
    public int openLock(String[] deadends, String target) {
        Set<String> seen = new HashSet<>();
        seen.add("0000");
        
        Set<String> dead = new HashSet<>();
        for(String d: deadends){
            dead.add(d);
        }
        
        Queue<String> q = new LinkedList<>();
        q.offer("0000");
        Queue<Integer> m = new LinkedList<>();
        m.offer(0);
        
        while(!q.isEmpty()){
            String node = q.poll();
            int move = m.poll();
            
            if(node.equals(target)){
                return move;
            }
            
            if(dead.contains(node)){
                continue;  
            }
            
            for(String nei: neighbors(node)){
                if(!seen.contains(nei)){
                    // System.out.println(nei);
                    seen.add(nei);
                    q.offer(nei);
                    m.offer(move+1);
                }
            }
            
        }
        
        return -1;
        
    }
    
    private String[] neighbors(String node){
        String[] neis = new String[8];
        int c = 0;
        
        for(int i=0; i<4; i++){
            int x = node.charAt(i)-'0';
            for(int j=-1; j<=1; j+=2){
                int y = (x+j+10)%10;
                neis[c] = node.substring(0,i)+String.valueOf(y)+node.substring(i+1);
                c++;
            }
        }
        return neis;
    }
}